# Data Cleaning

In [1]:
import numpy as np
import pandas as pd
df = pd.read_csv("Carbon_Projects.csv")
print(df.head)

<bound method NDFrame.head of         ID                                               Name  \
0     5302       SD Vista - Solar Water Pump Project in Kenya   
1     5264                 Yunfu LFG Power Generation Project   
2     5214  Pingjiang County Domestic Waste Harmless Landf...   
3     5203  Luoding BCCY New Power CO., Ltd. MSW biogas to...   
4     5186              Quanzhou Canhua PET Recycling Project   
...    ...                                                ...   
1951  2166  Chongqing Youyang County Youchou Hydropower St...   
1952  2162  2 x 3.5 MW Ullunkal Hydro Power Project in Ker...   
1953  2157                                    BAESA Project `   
1954  2136                     cancelled duplicate of VCSR218   
1955  2126  7.3 MW Bundled Wind Power Project by Oswal Cables   

                                              Proponent  \
0                              Sunculture Kenya Limited   
1                                   Multiple Proponents   
2           

In [2]:
# The data set dimension is 1956 x 13.
# Now, we want to see if there are any missing values.
print(df.isnull().any())

ID                                      False
Name                                     True
Proponent                               False
Project Type                            False
AFOLU Activities                         True
Methodology                             False
Status                                  False
Country/Area                            False
Estimated Annual Emission Reductions    False
Region                                   True
Project Registration Date                True
Crediting Period Start Date              True
Crediting Period End Date                True
dtype: bool


In [3]:
# We want to check the data type of each column.
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1956 entries, 0 to 1955
Data columns (total 13 columns):
 #   Column                                Non-Null Count  Dtype 
---  ------                                --------------  ----- 
 0   ID                                    1956 non-null   int64 
 1   Name                                  1955 non-null   object
 2   Proponent                             1956 non-null   object
 3   Project Type                          1956 non-null   object
 4   AFOLU Activities                      533 non-null    object
 5   Methodology                           1956 non-null   object
 6   Status                                1956 non-null   object
 7   Country/Area                          1956 non-null   object
 8   Estimated Annual Emission Reductions  1956 non-null   object
 9   Region                                1875 non-null   object
 10  Project Registration Date             1553 non-null   object
 11  Crediting Period Start Date   

In [4]:
# To change the data type of column Estimated Annual Emission Reduction to integer for further analysis
df['Estimated Annual Emission Reductions'] = df['Estimated Annual Emission Reductions'].str.replace(',', '', regex=True).astype('Int64')

In [5]:
# To confirm the data type of the column as integer64
print(df['Estimated Annual Emission Reductions'].dtype)

Int64


In [6]:
print(df['Estimated Annual Emission Reductions'])

0        40000
1        47759
2        60882
3        44555
4        58574
         ...  
1951    334000
1952     16125
1953    318793
1954    115912
1955     14832
Name: Estimated Annual Emission Reductions, Length: 1956, dtype: Int64


In [ ]:
# We know that there is no missing value in Project Type column.
# Now, we want to look at the carbon projects by project type.
unique_count = df['Project Type'].unique()

In [9]:
unique_count

array(['Energy distribution',
       'Energy industries (renewable/non-renewable sources); Waste handling and disposal',
       'Waste handling and disposal', 'Transport',
       'Energy industries (renewable/non-renewable sources); Mining/mineral production',
       'Agriculture Forestry and Other Land Use', 'Energy demand',
       'Livestock, enteric fermentation, and manure management; Waste handling and disposal',
       'Energy industries (renewable/non-renewable sources); Fugitive emissions from fuels (solid, oil and gas); Mining/mineral production',
       'Energy industries (renewable/non-renewable sources); Livestock, enteric fermentation, and manure management; Waste handling and disposal',
       'Energy demand; Waste handling and disposal',
       'Energy industries (renewable/non-renewable sources); Transport',
       'Livestock, enteric fermentation, and manure management',
       'Energy industries (renewable/non-renewable sources)',
       'Transport; Waste handling and

In [13]:
df['Project Type'].agg(['count', 'nunique'])

count      1956
nunique      32
Name: Project Type, dtype: int64

In [15]:
df['Project Type'].value_counts()

Project Type
Energy industries (renewable/non-renewable sources)                                                                                         745
Agriculture Forestry and Other Land Use                                                                                                     536
Energy demand                                                                                                                               246
Energy industries (renewable/non-renewable sources); Waste handling and disposal                                                            159
Waste handling and disposal                                                                                                                  63
Energy industries (renewable/non-renewable sources); Livestock, enteric fermentation, and manure management; Waste handling and disposal     59
Livestock, enteric fermentation, and manure management; Waste handling and disposal                                        

In this dataset, there are 32 unique project types.

In the Project Type column, the project types registered in Verra are:

- Energy Distribution
- Energy industries (renewable/non-renewable): waste handlind and disposal
- Wate Handling and disposal
- Transport
- 

In [20]:
print(df['Status'].unique())

['Withdrawn' 'Registered' 'Rejected by Administrator'
 'Registration and verification approval request denied'
 'Registration request denied']


In [60]:
df['Status'].agg(['count', 'unique', 'nunique'])

count                                                   1956
unique     [Withdrawn, Registered, Rejected by Administra...
nunique                                                    5
Name: Status, dtype: object

In [29]:
df['Status'].value_counts()

Status
Registered                                               1461
Withdrawn                                                 240
Rejected by Administrator                                 161
Registration and verification approval request denied      79
Registration request denied                                15
Name: count, dtype: int64

As I perform value_count function from pandas library on `Status` column, the dataset has:
- Registered: 1461
- Withdrawn: 240
- Rejected by Administrator: 161
- Registration and verification approval request denied: 79
Registration request denied: 15

In [31]:
Approval_percentage = print(1461/1956)

0.7469325153374233


In [64]:
df['Status'].value_counts('Registered')

Status
Registered                                               0.746933
Withdrawn                                                0.122699
Rejected by Administrator                                0.082311
Registration and verification approval request denied    0.040389
Registration request denied                              0.007669
Name: proportion, dtype: float64

In [65]:
(df['Status'] == 'Registered').sum()

np.int64(1461)

74.69 percent of projects in Verra are registered, and not denied.

In [ ]:
# To confirm the shape of the 
print(f"Registered projects: {df_registered.shape[0]}")
print(f"Not Registered projects: {df_not_registered.shape[0]}")

Registered projects: 1461
Not Registered projects: 495


In [ ]:
# To convert filtered datasets into separate CSV files
df_registered.to_csv("Registered_Projects.csv", index=False)
df_not_registered.to_csv("Not_Registered_Projects.csv", index=False)

In [ ]:
df['Methodology'].agg(['count', 'unique', 'nunique'])
df['Methodology'].value_counts()
# We wee that ACM0002 is the most common methodology

Methodology
ACM0002               484
VMR0006               145
AMS-I.D.              142
AR-ACM0003            127
AMS-III.AU            111
                     ... 
VM0004                  1
VM0001                  1
VM0025; VMD0038         1
AMS-I.C.; AMS-I.D.      1
AMS-III.Y               1
Name: count, Length: 140, dtype: int64

Energy (renewable/non-renewable)
Energy distribution
Energy demand
Manufacturing industries
Chemical industry
Construction
Transport
Mining/Mineral production
Metal production
Fugitive emissions – from fuels (solid, oil, and gas)
Fugitive emissions – from Industrial gases (halocarbons and sulphur hexafluoride)
Solvents use
Waste handling and disposal
Agriculture, forestry and other land use (AFOLU)
Livestock and manure management
Carbon capture and storage

In [63]:
labels = ['Energy (renewable/non-renewable)',
          'Energy distribution',
          'Energy demand',
          'Manufacturing industries',
          'Chemical industry',
          'Construction',
          'Transport',
          'Mining/Mineral production',
          'Metal production',
          'Fugitive emissions – from fuels (solid, oil, and gas)',
          'Fugitive emissions – from Industrial gases (halocarbons and sulphur hexafluoride)',
          'Solvents use',
          'Waste handling and disposal',
          'Agriculture, forestry and other land use (AFOLU)',
          'Livestock and manure management',
          'Carbon capture and storage']

Labels_Data = pd.DataFrame(labels)

In [66]:
df['Project Type'].value_counts()

Project Type
Energy industries (renewable/non-renewable sources)                                                                                         745
Agriculture Forestry and Other Land Use                                                                                                     536
Energy demand                                                                                                                               246
Energy industries (renewable/non-renewable sources); Waste handling and disposal                                                            159
Waste handling and disposal                                                                                                                  63
Energy industries (renewable/non-renewable sources); Livestock, enteric fermentation, and manure management; Waste handling and disposal     59
Livestock, enteric fermentation, and manure management; Waste handling and disposal                                        

In [67]:
Registered = pd.read_csv("Registered_Projects.csv")

In [68]:
Registered['Project Type'].value_counts()

Project Type
Energy industries (renewable/non-renewable sources)                                                                                         664
Agriculture Forestry and Other Land Use                                                                                                     226
Energy demand                                                                                                                               210
Energy industries (renewable/non-renewable sources); Waste handling and disposal                                                            139
Waste handling and disposal                                                                                                                  57
Energy industries (renewable/non-renewable sources); Livestock, enteric fermentation, and manure management; Waste handling and disposal     34
Livestock, enteric fermentation, and manure management; Waste handling and disposal                                        

In [ ]:
Registered.isnull().sum()


ID                                         0
Name                                       1
Proponent                                  0
Project Type                               0
AFOLU Activities                        1423
Methodology                                0
Status                                     0
Country/Area                               0
Estimated Annual Emission Reductions       0
Region                                    81
Project Registration Date                403
Crediting Period Start Date              508
Crediting Period End Date                508
dtype: int64

In [75]:
df.isnull().sum()

ID                                         0
Name                                       1
Proponent                                  0
Project Type                               0
AFOLU Activities                        1423
Methodology                                0
Status                                     0
Country/Area                               0
Estimated Annual Emission Reductions       0
Region                                    81
Project Registration Date                403
Crediting Period Start Date              508
Crediting Period End Date                508
dtype: int64

Registered['Region'].isnull()

In [ ]:
missing_region_Registered_df = df[df['Country/Area'].notnull() & df['Region'].isnull()]

print(missing_region_Registered_df['Country/Area'])


1               China
2               China
5               China
14              China
18              China
24              China
34              China
109             China
425            Turkey
428            Turkey
522             China
524             China
582           Nigeria
614             China
699             India
789        Madagascar
1092            India
1134            China
1147            China
1156      Netherlands
1159           Turkey
1167           Turkey
1172            India
1197           Turkey
1234            China
1288    United States
1290           Turkey
1291           Turkey
1299            China
1306         Thailand
1307            China
1309            China
1317            China
1324            India
1329            China
1331            China
1333            India
1334            China
1335            China
1337            China
1338            China
1339            China
1340            China
1341            India
1365            India
1373      

In [83]:
print(missing_region_Registered_df['Country/Area'].value_counts())

Country/Area
China            38
India            19
Turkey            9
Thailand          4
Brazil            3
Colombia          2
Nigeria           1
Madagascar        1
United States     1
Netherlands       1
Senegal           1
Pakistan          1
Name: count, dtype: int64


In [86]:
Missing_Region1_Labels = {'China' : 'Asia',
                         'India': 'Asia',
                         'Turkey': 'Asia',
                         'Thailand': 'Asia',
                         'Brazil': 'Latin America',
                         'Colombia': 'Latin America',
                         'Nigeria': 'Africa',
                         'Madagascar': 'Africa',
                         'United States': 'North America',
                         'Netherlands': 'Europe',
                         'Senegal': 'Africa',
                         'Pakistan': 'Asia'}



In [87]:
Missing_Region_labels = {
    country: 'Asia' for country in ['China', 'India', 'Pakistan', 'Turkey', 'Thailand']
 } | {
     country: 'Latin America' for country in ['Brazil', 'Colombia']
 } | {
     country: 'Africa' for country in ['Nigeria', 'Madagascar', 'Senegal']
 } | {
     'United States': 'North America',
     'Netherlands': 'Europe'
 }

In [88]:
df.loc[df['Region'].isna() & df['Country/Area'].isin(Missing_Region1_Labels),
       'Region'
       ] = df['Country/Area'].map(Missing_Region1_Labels)

In [89]:
# To convert filtered datasets into separate CSV files
df.to_csv("Carbon_Projects_Version2.csv", index=False)

In [90]:
# We want to filter data by the project status: registered or not
df_registered = df[df['Status'] == 'Registered']
df_not_registered = df[df['Status'] != 'Registered']

In [91]:
# To convert filtered datasets into separate CSV files
df_registered.to_csv("Registered_Projects.csv", index=False)
df_not_registered.to_csv("Not_Registered_Projects.csv", index=False)

In [4]:
import pandas as pd
Registered = pd.read_csv("Carbon_Projects_Version2.csv")

Registered['Country/Area'].value_counts()


Country/Area
China         630
India         460
Turkey        154
Brazil        101
Colombia       49
             ... 
Ecuador         1
Costa Rica      1
Ethiopia        1
Georgia         1
Tajikistan      1
Name: count, Length: 102, dtype: int64

In [5]:
# Let us make a dictionary that matches countries/areas to proper regions


country_region_map = {
    'Albania': 'Southern Europe',
    'Angola': ' Middle Africa',
    'Argentina': 'South America',
    'Austria': 'Western Europe',
    'Bahamas': 'Central America and the Caribbean',
    'Bahrain': 'Western Asia',
    'Bangladesh': 'Southern Asia',
    'Belize': 'Central America',
    'Benin': 'Western Africa',
    'Brazil': 'South America',
    'Burkina Faso': 'Western Africa',
    'Burundi': 'Eastern Africa',
    'Cambodia': 'South-eastern Asia',
    'Cameroon': 'Middle Africa',
    'Canada': 'North America',
    'Central African Republic': 'Middle Africa',
    'Chad': 'Middle Africa',
    'Chile': 'South America',
    'China': 'Eastern Asia',
    'Colombia': 'South America',
    'Comoros': 'Eastern Africa',
    'Congo': 'Middle Africa',
    'Congo, The Democratic Republic of The': 'Middle Africa',
    'Costa Rica': 'Central America',
    'Cote D\'Ivoire': 'Africa',
    'Dominican Republic': 'Caribbean',
    'Ecuador': 'South America',
    'Egypt': 'Northern Africa',
    'El Salvador': 'Central America',
    'Estonia': 'Northern Europe',
    'Ethiopia': 'Eastern Africa',
    'Fiji': 'Melanesia',
    'Gambia': 'Western Africa',
    'Germany': 'Western Europe',
    'Ghana': 'Western Africa',
    'Guatemala': 'Central America',
    'Guinea-Bissau': 'Western Africa',
    'Honduras': 'Central America',
    'Iceland': 'Northern Europe',
    'India': 'Southern Asia',
    'Indonesia': 'South-eastern Asia',
    'Israel': 'Western Asia',
    'Italy': 'Southern Europe',
    'Japan': 'Eastern Asia',
    'Jordan': 'Western Asia',
    'Kazakhstan': 'Central Asia',
    'Kenya': 'Eastern Africa',
    'Lao': 'South-eastern Asia',
    'Liberia': 'Western Africa',
    'Lithuania': 'Northern Europe',
    'Madagascar': 'Eastern Africa',
    'Malawi': 'Eastern Africa',
    'Malaysia': 'South-eastern Africa',
    'Mauritania': 'Western Africa',
    'Mauritius': 'Eastern Africa',
    'Mexico': 'Central America',
    'Morocco': 'Northern Africa',
    'Mozambique': 'Eastern Africa',
    'Myanmar': 'South-eastern Asia',
    'Namibia': 'Southern Africa',
    'Nepal': 'Southern Asia',
    'Netherlands': 'Western Europe',
    'New Zealand': 'Australia and New Zealand',
    'Nicaragua': 'Central America',
    'Niger': 'Western Africa',
    'Nigeria': 'Western Africa',
    'Oman': 'Western Asia',
    'Pakistan': 'Southern Asia',
    'Panama': 'Central America',
    'Papua New Guinea': 'Melanesia',
    'Paraguay': 'South America',
    'Peru': 'South America',
    'Philippines': 'South-eastern Asia',
    'Romania': 'Eastern Europe',
    'Russian Federation': 'Eastern Europe',
    'Rwanda': 'Eastern Africa',
    'Saudi Arabia': 'Western Asia',
    'Senegal': 'Western Africa',
    'Serbia': 'Southern Europe',
    'Sierra Leone': 'Western Africa',
    'Singapore': 'South-eastern Asia',
    'South Africa': 'Southern Africa',
    'South Korea': 'Eastern Asia',
    'Spain': 'Southern Europe',
    'Sri Lanka': 'Southern Asia',
    'Switzerland': 'Western Europe',
    'Tajikistan': 'Central Asia',
    'Tanzania': 'Easetern Africa',
    'Thailand': 'South-eastern Asia',
    'Timor-Leste': 'Indonesia',
    'Togo': 'Western Africa',
    'Turkey': 'Western Asia',
    'Uganda': 'Eastern Africa',
    'United Kingdom': 'Northern Europe',
    'United States': 'Northern America',
    'Uruguay': 'South America',
    'Uzbekistan': 'Central Asia',
    'Viet Nam': 'South-eastern Asia',
    'Zambia': 'Eastern Africa',
    'Zimbabwe': 'Eastern Africa'
    }

In [ ]:
# We want to first fit the dictionary to the dataset to ensure consistency



In [ ]:
# Then, we make a new dictionary:

